casper/src/main/scala/coop/rchain/casper/safety/CliqueOracle.scala

```diff
@@ -0,0 +1,178 @@
+package coop.rchain.casper.safety
+
+import cats.data.OptionT
+import cats.effect.{Concurrent, Sync}
+import cats.syntax.all._
+import coop.rchain.blockstorage.dag.BlockDagRepresentation
+import coop.rchain.blockstorage.syntax._
+import coop.rchain.casper.SafetyOracle
+import coop.rchain.casper.util.Clique
+import coop.rchain.metrics.Metrics.Source
+import coop.rchain.metrics.implicits._
+import coop.rchain.metrics.{Metrics, Span}
+import coop.rchain.models.BlockHash.BlockHash
+import coop.rchain.models.Validator.Validator
+import coop.rchain.models.syntax._
+import coop.rchain.shared.Log
+import coop.rchain.shared.syntax._
+import fs2.Stream
+
+object CliqueOracle {
+
+  private type M         = BlockHash    // type for message
+  private type V         = Validator    // type for message creator
+  private type WeightMap = Map[V, Long] // stakes per message creator
+
+  implicit val baseMetricsSource: Source = coop.rchain.casper.CasperMetricsSource
+
+  /** weight map of main parent (fallbacks to message itself if no parents)
+     TODO - why not use local weight map but seek for parent?
+    P.S. This is related to the fact that we create latest message for newly bonded validator
+    equal to message where bonding deploy has been submitted. So stake from validator that did not create anything is
+    put behind this message. So here is one more place where this logic makes things more complex.*/
+  private def getCorrespondingWeightMap[F[_]: Sync](
+      target: M,
+      dag: BlockDagRepresentation[F]
+  ): F[WeightMap] =
+    dag.lookupUnsafe(target).flatMap { meta =>
+      meta.parents.headOption match {
+        case Some(mainParent) => dag.lookupUnsafe(mainParent).map(_.weightMap)
+        case None             => meta.weightMap.pure
+      }
+    }
+
+  /**
+    * If two validators will never have disagreement on target message
+    *
+    * Prerequisite for this is that latest messages from a and b both are in main chain with target message
+    *
+    *     a    b
+    *     *    *  <- lmB
+    *      \   *
+    *       \  *
+    *        \ *
+    *         \*
+    *          *  <- lmAjB
+    *
+    * 1. get justification of validator b as per latest message of a (lmAjB)
+    * 2. check if any self justifications between latest message of b (lmB) and lmAjB are NOT in main chain
+    *    with target message.
+    *
+    *    If one found - this is a source of disagreement.
+    * */
+  private def neverEventuallySeeDisagreement[F[_]: Sync](
+      validatorA: V,
+      validatorB: V,
+      dag: BlockDagRepresentation[F],
+      targetMessage: M
+  ): F[Boolean] = {
+    def mightEventuallyDisagree(lmB: BlockHash, lmAjB: BlockHash): F[Boolean] =
```

Comment:
This is a typo - I'd like to use M here as well. As it stands for message. Thats why I don't want to have BlockHash everywhere. But as BlockDagRepresentation needs BlockHash, I defined type alias.

Reply:
Why not call it `Message` or `Msg` instead of `M`?

The idea of an alias is also to make it easier for the reader to follow the code.

Reply:
It is said that this is a message here https://github.com/rchain/rchain/blob/bf23fb1ee2492f422fe6243fa04f9f96ecd9a379/casper/src/main/scala/coop/rchain/casper/safety/CliqueOracle.scala#L22
And pretty often you have to pass in variables like this 
https://github.com/rchain/rchain/blob/bf23fb1ee2492f422fe6243fa04f9f96ecd9a379/casper/src/main/scala/coop/rchain/casper/safety/CliqueOracle.scala#L67
So you duplicating name of the variable and type, which I want to avoid. As variable name is something that is used on code. But anyway, if you think this is a bad practice - I will adjust.

Reply:
I don't see duplication if existing alias is used. I'd expect that `M` is generic type.
```scala
targetMessage: BlockHash
```
If you like we can experiment on a more places to see if gives more clarity.